src/dataflow/src/render/reduce.rs

```diff
@@ -275,242 +291,324 @@ where
-                } else {
-                    // We respect the multiplicity here (unlike in hierarchical aggregation)
-                    // because we don't know that the aggregation method is not sensitive
-                    // to the number of records.
-                    let iter = source.iter().flat_map(|(v, w)| {
-                        std::iter::repeat(v.iter().next().unwrap()).take(*w as usize)
-                    });
-                    if prepend_key {
-                        row_packer.extend(key.iter());
+    // Perform a final aggregation, on potentially hierarchically reduced data.
+    // The same code should work on data that can not be hierarchically reduced.
+    partial.reduce_abelian::<_, OrdValSpine<_, _, _, _>>("ReduceInaccumulable", {
+        let mut row_packer = RowPacker::new();
+        move |key, source, target| {
+            // Negative counts would be surprising, but until we are 100% certain we wont
+            // see them, we should report when we do. We may want to bake even more info
+            // in here in the future.
+            if source.iter().any(|(_val, cnt)| cnt < &0) {
+                // XXX: This reports user data, which we perhaps should not do!
+                for (val, cnt) in source.iter() {
+                    if cnt < &0 {
+                        log::error!("[customer-data] Negative accumulation in ReduceInaccumulable: {:?} with count {:?}", val, cnt);
                     }
-                    row_packer.push(func.eval(iter, &RowArena::new()));
-                    target.push((row_packer.finish_and_reuse(), 1));
                 }
+            } else {
+                // We respect the multiplicity here (unlike in hierarchical aggregation)
+                // because we don't know that the aggregation method is not sensitive
+                // to the number of records.
+                let iter = source.iter().flat_map(|(v, w)| {
+                    std::iter::repeat(v.iter().next().unwrap()).take(*w as usize)
+                });
+                if prepend_key {
+                    row_packer.extend(key.iter());
+                }
+                row_packer.push(func.eval(iter, &RowArena::new()));
+                target.push((row_packer.finish_and_reuse(), 1));
             }
-        })
-    };
-
-    ok_out
+        }
+    })
 }
 
-/// Builds the dataflow for a reduction that can be performed in-place.
+/// Builds the dataflow for a reductions that can be performed in-place.
 ///
 /// The incoming values are moved to the update's "difference" field, at which point
 /// they can be accumulated in place. The `count` operator promotes the accumulated
 /// values to data, at which point a final map applies operator-specific logic to
 /// yield the final aggregate.
-fn build_accumulable<G>(
+///
+/// If `prepend_key` is specified, the key is prepended to the arranged values, making
+/// the arrangement suitable for publication itself.
+fn build_accumulables<G>(
```

Comment:
the implementation allocates diff vectors that are wide enough to cover the max index of accumulable aggregates (in the original list of aggregations). That ends up being both an overallocation and kind of confusing. Instead, I think we could change the behavior to allocate diff vectors that are only 3 * number of accumulable indexes wide

Reply:
If this is the case, it seems out of sync with the code that peels the accumulable values off of the accumulation, which just goes one by one. That sounds like a correctness issue if so (if they are produced sparse here, and then consumed if as dense up above in the finishing reduction).

Reply:
so at the end of this function we stuff the final values for the aggregations we care about into a row, so the later code which does the peeling one by one is fine.

I think we can just simplify this function by passing a vec of aggregatefuncs (instead of pairs of func and index) and then keeping a vector of diffs for only the aggregates we care about kind of falls out naturally

Reply:
Yeah, that seems right, sorry. Cleaning it up would be great, once it is in a state that you are comfortable with the correctness!

Reply:
I think its now cleaned up!